lib/zstd.h

```diff
@@ -1621,6 +1625,59 @@ ZSTDLIB_API size_t ZSTD_CCtx_refPrefix_advanced(ZSTD_CCtx* cctx, const void* pre
  */
 #define ZSTD_c_enableDedicatedDictSearch ZSTD_c_experimentalParam8
 
+/* ZSTD_c_stableInBuffer
+ * Experimental parameter.
+ * Default is 0 == disabled. Set to 1 to enable.
+ *
+ * Tells the compressor that the ZSTD_inBuffer will ALWAYS be the same
+ * between calls, except for the modifications that zstd makes to pos (the
+ * caller must not modify pos). This is checked by the compressor, and
+ * compression will fail if it ever changes. This means the only flush
+ * mode that makes sense is ZSTD_e_end, so zstd will error if ZSTD_e_end
+ * is not used. The data in the ZSTD_inBuffer in the range [src, src + pos)
+ * MUST not be modified during compression or you will get data corruption.
+ *
+ * When this flag is enabled zstd won't allocate an input window buffer,
+ * because the user guarantees it can reference the ZSTD_inBuffer until
+ * the frame is complete. But, it will still allocate an output buffer
+ * large enough to fit a block (see ZSTD_c_stableOutBuffer). This will also
+ * avoid the memcpy() from the input buffer to the input window buffer.
+ *
+ * NOTE: ZSTD_compressStream2() will error if ZSTD_e_end is not used.
+ * That means this flag cannot be used with ZSTD_compressStream().
+ *
+ * NOTE: So long as the ZSTD_inBuffer always points to valid memory, using
+ * this flag is ALWAYS memory safe, and will never access out-of-bounds
+ * memory. However, compression WILL fail if you violate the preconditions.
+ *
+ * WARNING: The data in the ZSTD_inBuffer in the range [dst, dst + pos) MUST
+ * not be modified during compression or you will get data corruption. This
+ * is because zstd needs to reference data in the ZSTD_inBuffer to find
+ * matches. Normally zstd maintains its own window buffer for this purpose,
+ * but passing this flag tells zstd to use the user provided buffer.
+ */
+#define ZSTD_c_stableInBuffer ZSTD_c_experimentalParam9
+
+/* ZSTD_c_stableOutBuffer
```

Comment:
I was reading the definition of this parameter,
and considering : is it really about being "stable" ?

In contrast with the input buffer, it doesn't really matter that the output buffer is not flushed between invocations.
It can. Which means, the output buffer doesn't have to be "stable".

What matters is the pledge that the output buffer should be "large enough" to not require an intermediate buffer.

In which case, maybe the name of the parameter could reflect that property instead.

Reply:
> What matters is the pledge that the output buffer should be "large enough" to not require an intermediate buffer.

Thats not quite it either. For correctness, it either:

* Has enough space remaining to not require an intermediate buffer
* Or the space remaining never increases and the user is okay with the `dstSizeTooSmall` error (so if `dstSizeTooSmall` is returned, it will always be returned).

I've opted to require stability, because that is a simpler contract. But we could loosen the requirements.

If you want to loosen the requirements, what do you think a good parameter name would be?

Reply:
It's not clear to me how one is supposed to use ZSTD_c_stableOutBuffer with streaming compression, since the output buffer has to be flushed eventually, and hence output.pos reset to 0. Am I missing something?

Reply:
@naav `ZSTD_c_stableOutBuffer` is not useful for general streaming compression. It is useful in two scenarios:

1. Implementing a non-streaming interface optimally using the streaming compression interface. E.g. `ZSTD_compress2()` is implemented using `ZSTD_compressStream2()` together with `ZSTD_c_stableInBuffer` and `ZSTD_c_stableOutBuffer`.
2. It can also be useful if you're streaming the input buffer, but not streaming the output buffer. E.g. the output buffer is `mmaped()`. Though this might not be particularly useful, since the more useful buffer to mmap would be the input buffer.